### PR TITLE
Use both fetch/push url for build_tarball

### PR DESCRIPTION
**Pull request by Paul-Ferrell**
Currently build_tarball only checks for the existence of a binary dist
at the push url. We need it to check both the push and fetch url's,
so that we only rebuild parts of the binary cache that don't exist in
our remote cache.

### CODE REVIEW

**Comment by becker33** [comment]
@scottwittenburg can you take a look at this? It looks reasonable to me, but you're more likely than I am to see an issue with it.

File: lib/spack/spack/binary_distribution.py
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py]: I might be getting confused by the variable names, why is this called `local_spackfile_path`?  It seems the "remote" url is coming from the mirrors `fetch_url` while the "local" one is coming from the `push_url`. 

@Paul-Ferrell  You aren't thinking of these as separate mirrors, are you?  The idea with allowing mirrors to have separate fetch and push urls is to handle situations where you might want to have your  S3 mirror behind an http proxy and you couldn't push things to a brainless http mirror.

File: lib/spack/spack/binary_distribution.py
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py]: I think it's weird to be possibly calling `remove_url` on what should be the same tarball twice.  If you remove it via the `local_spackfile_path` (the `push_url` version), it should also no longer be present to be removed in the block below.

**Review by scottwittenburg — CHANGES_REQUESTED** [review]
If I understand the problem correctly, and I might not, this doesn't seem the way to solve it.  I think better would be to change the code to only ever try to read from the `fetch_url` and only ever try to write to the `push_url`.   Which seems like something we should have been doing all along, but maybe haven't run into it yet because not many people are using the distinct urls for fetching and pushing. 

So if the use case for this is really two different ways to address the same mirror, then I think the right fix is to use the `fetch_url` to see if the tarball exists, and then use the `push_url` both to remove the existing tarball as well as to write the new one.

It's possible I'm way off base here, @opadron came up with the distinct urls approach and might see things differently.  Perhaps he can provide some insight here.

File: lib/spack/spack/binary_distribution.py
Comment (Paul-Ferrell) [reply lib/spack/spack/binary_distribution.py]: I changed those variable names to make things more clear (push_ and fetch_ vs local_ and remote_). 

I've made another slight change. Now if the fetch url exists 'force' will simply ignore it rather than trying to delete it, which in many cases Spack can't do anyway.  So now no 'writing to'/'deleting from' the fetch url.   

I'm not thinking of these as separate mirrors, but as the same mirror connected by an external mechanism or with different paths for security purposes. In my case, the mirror is just a simple web server; we can't use S3. Like the S3 case, pushing is a separate path with separate security mechanisms. Unlike the S3 case, the push to the file server is done with a separate process over client cert authenticated HTTP PUT requests. I guess I could add support for that in Spack, but it seems odd to me.

File: lib/spack/spack/binary_distribution.py
Comment (Paul-Ferrell) [reply lib/spack/spack/binary_distribution.py]: @scottwittenburg 
I agree that we shouldn't be trying to write to or delete from the fetch url at all.  See above.

File: lib/spack/spack/binary_distribution.py
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py]: Before this change, the `outdir` passed into this method was created like:

```
outdir = url_util.format(mirror.push_url)
```

I'm not sure it's critical, but to me it seems there's no reason to change it arbitrarily.  Let me know if there's something I'm missing.

File: lib/spack/spack/binary_distribution.py
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py:946]: Can't it be done like:

```
if web_util.url_exists(fetch_spackfile_path):
    web_util.remove_url(push_spackfile_path)
```

I left out the force bit force bit for brevity.  This is what I imagined in my previous review: Since those should be the same file, read from the fetch url, write to the push url.

File: lib/spack/spack/binary_distribution.py
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py]: I was thinking if you did like I suggested above, you can then remove this second block checking the url existence a second time.

File: lib/spack/spack/binary_distribution.py
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py]: Then here pushing would be done with the push url instead of the fetch one.

**Review by scottwittenburg — CHANGES_REQUESTED** [review]
Thanks for clarifying your use case @Paul-Ferrell it seems like you're using the fetch/push urls as intended, and we're on the same page there.  

I think down at the bottom you probably meant to change the `push_to_url` call to use the push url instead of the fetch one, so I marked that.  I also suggested what I think would be the right way to handle the existence check and removal (with one `if` block rather than two), but let me know if you disagree.

File: lib/spack/spack/binary_distribution.py
Comment (Paul-Ferrell) [reply lib/spack/spack/binary_distribution.py]: You're right, I'll fix this.

File: lib/spack/spack/binary_distribution.py
Comment (Paul-Ferrell) [reply lib/spack/spack/binary_distribution.py:946]: From what you've said, push/fetch feature was designed around the idea that while the URL's are different, they are effectively the same place. I think they don't necessarily have to be, and am in fact counting on that. By letting them be two distinct unconnected spaces, it allows you to set up a mirror such that buildcache will only create binary cached builds you don't already have at a given mirror, and storing them wherever you want them to go. 

In our case, we then push those builds to that same mirror server, but that doesn't necessarily have to be the next step.

This is a bit of a concept change with this feature, I guess. It means you can't assume that a file at one url exists at the other, or that deleting from the push location deletes the file at the fetch location. This CAN be true, but doesn't have to be.

File: lib/spack/spack/binary_distribution.py
Comment (Paul-Ferrell) [reply lib/spack/spack/binary_distribution.py]: That's just a mistake. I thought I checked to make sure remote_spackfile_path wasn't used anywhere else. I missed this one.

**Comment by Paul-Ferrell** [comment]
I went ahead and applied the same basic logic to how the spec files are pushed, and fixed the issue with the final upload step.

File: lib/spack/spack/binary_distribution.py
Comment (opadron) [reply lib/spack/spack/binary_distribution.py:946]: There's a design assumption that the two urls present a consistent view of the same build cache.  You are correct that we don't currently do anything in the spack code that strictly requires that they do, but that could change.

At a minimum, if spack pushes a binary to the build cache via the push url, it's expected that the same binary would then be found at the fetch url, and if something is deleted via the push url, it would no longer be found under the fetch url.

Comment (opadron) [reply lib/spack/spack/binary_distribution.py:946]: I should further clarify: what I wrote above is only the intended semantics.  You can certainly point the two urls at completely different locations, but users should do so knowing that spack may treat those locations as consistent.  If they're not consistent, this will produce unintended side effects, and while in this example use case, you actually wanted those unintended effects, we can't support that case as it would likely make future improvements that would strictly require consistency impossible.

Your use case is better modeled by two different build caches, rather than a single cache with different urls.  And your desired behavior sounds like some kind of new feature where Spack would use one cache as a mirror or caching layer for another.  I would enthusiastically support something like this, but not with the current implementation, as it relies on an unintended feature of push and fetch urls, and I don't think we should be formally supporting that feature.

**Review by opadron — DISMISSED** [review]
@Paul-Ferrell and I went through the details of this PR over Slack, and now I understand the idea. I was under the mistaken impression that the changes induced Spack behaviour that relied on the push and fetch urls being different, but that is not the case.

**Comment by Paul-Ferrell** [comment]
@scottwittenburg I think this is just waiting on your approval.

**Comment by scottwittenburg** [comment]
This PR violates my mental model of how mirrors work and the intention of fetch vs push urls for mirrors.  I agree it now does the extra check of the fetch url, which you stated as the goal in your PR description, but for most cases people have a single url for fetch/push, and now we're duplicating every check we did before twice (two checks of the `spec.yaml`, two checks of the `.spack` file.


I think I need more time to think about the issue before approving this.  I may need to reach for additional information about your use case as well.

**Comment by Paul-Ferrell** [comment]
@scottwittenburg 
Yeah, I basically told my team leader: "I'm fixing this problem with a simple change that turns an invalid configuration into a valid one, but it breaks the Spack dev's mental model of how things work so it's an uphill climb getting it through".

So there are effectively three cases here:
 
1. `push == fetch` - No change in how things work (especially with my last push, which skips the second file existence check in this case).
2. `push != fetch`, but through remote mechanisms they are effectively the same location - Slight change in that we check both push and fetch urls for file existence, but otherwise no change.
3. `push != fetch`, with no guarantees that the locations are synced in any way - This was an unsupported configuration. It wasn't an error (there's no way to generically validate that you have case 2 vs 3), but it didn't work.

Enabling #3 does two things, both of which we (LANL TCE Project) need: 

1. It allows you to build a binary cache without rebuilding packages that already exist in your remote binary cache. This allows us to build a large environment in CI without having to have the binary cache 'preloaded' locally. The S3 scheme is great, but we can't use it in our environment. Another potential solution is to give two mirrors for `create buildcache`, one to filter against and one to build to, but I think that unnecessarily complicates things more than the proposed solution.
2. It allows you to utilize external syncing mechanisms between the push and fetch locations. There's no true standard way to sync files to remote URL's. While it might make sense to put a few popular mechanisms in place, there are a wide variety of web API's and protocols that people might need to use beyond what you internally support. In our case, we're using fairly standard (albeit rarely used) HTTP PUT with client cert authentication. I could (and would) add this as another available URL scheme, but even though it's part of the HTTP standard I still think it's esoteric. If I have to run my CI somewhere where I can't push via HTTP, but can via FTP or some other weirdness that hops that tunnels through multiple SSH hops, I need to be able to.

It does have one more side-effect: It breaks the implicit idea that deleting from the push url deletes from the fetch url. As Spack is, this is not guaranteed anyway unless the URL scheme supports it. As the 'URL schemes'/'remote file handling mechanisms' are all non-standard, there's no guarantee of this anyway except on a URL scheme by URL scheme basis.

**Comment by scottwittenburg** [comment]
> @scottwittenburg
> Yeah, I basically told my team leader: "I'm fixing this problem with a simple change that turns an invalid configuration into a valid one, but it breaks the Spack dev's mental model of how things work so it's an uphill climb getting it through".

Ok. 

> So there are effectively three cases here:
> 
> 1. `push == fetch` - No change in how things work (especially with my last push, which skips the second file existence check in this case).

I agree this is an improvement.

> 2. `push != fetch`, but through remote mechanisms they are effectively the same location - Slight change in that we check both push and fetch urls for file existence, but otherwise no change.
> 3. `push != fetch`, with no guarantees that the locations are synced in any way - This was an unsupported configuration. It wasn't an error (there's no way to generically validate that you have case 2 vs 3), but it didn't work.
> 
> Enabling #3 does two things, both of which we (LANL TCE Project) need:
> 
> 1. It allows you to build a binary cache without rebuilding packages that already exist in your remote binary cache. This allows us to build a large environment in CI without having to have the binary cache 'preloaded' locally. The S3 scheme is great, but we can't use it in our environment. Another potential solution is to give two mirrors for `create buildcache`, one to filter against and one to build to, but I think that unnecessarily complicates things more than the proposed solution.

This part I don't completely follow.  Since `spack buildcache create` is used to create a binary package from software you have already built with `spack install`, how does throwing an exception from `spack buildcache create` prevent you from rebuilding it unnecessarily?

> 2. It allows you to utilize external syncing mechanisms between the push and fetch locations. There's no true standard way to sync files to remote URL's. While it might make sense to put a few popular mechanisms in place, there are a wide variety of web API's and protocols that people might need to use beyond what you internally support. In our case, we're using fairly standard (albeit rarely used) HTTP PUT with client cert authentication. I could (and would) add this as another available URL scheme, but even though it's part of the HTTP standard I still think it's esoteric. If I have to run my CI somewhere where I can't push via HTTP, but can via FTP or some other weirdness that hops that tunnels through multiple SSH hops, I need to be able to.

Fair enough.

> It does have one more side-effect: It breaks the implicit idea that deleting from the push url deletes from the fetch url. As Spack is, this is not guaranteed anyway unless the URL scheme supports it. As the 'URL schemes'/'remote file handling mechanisms' are all non-standard, there's no guarantee of this anyway except on a URL scheme by URL scheme basis.

You've hit on one of the things that has bothered me about this.  I agree there were no guarantees about this before, and that the mirror fetch/push abstraction wasn't completely fleshed out   If we're going to address that in some way (which at first I thought this PR was attempting to do), we should be clear about what the model is.  If we accept this PR, then going forward we need to understand there is no implied correlation between the contents of a single mirror when it has different fetch vs push urls.  That could have implications in the future as there are other parts of the `buildcache` and `binary_distribution` modules (and probably others as well) that deal with reading from and writing to mirrors.  For example, because mirrors have so far been without intelligence, when we update a buildcache index, we read the spec.yaml files from the mirror and use that information to build an index and write it back to the mirror.   So far that code has not properly taken into account the fact we can have different fetch vs push urls, but when we go to fix that, what mental model should we have?  My inclination before this would have been simple: read the spec yamls from the fetch url, write the index back the to the push url, and it's up to the user to make sure those reflect the same version of reality.

I guess I could be ok with this simply for the sake of getting you unblocked, but I'm worried about potentially having to treat all mirrors as essentially two different mirrors where only one of the them is writable.

**Comment by Paul-Ferrell** [comment]
> This part I don't completely follow. Since spack buildcache create is used to create a binary package from software you have already built with spack install, how does throwing an exception from spack buildcache create prevent you from rebuilding it unnecessarily?

Because throwing an exception is how `build_tarball` function communicates to the `buildcache create` that something shouldn't be built. That's just how it was when I got here. I think the 'should build' decision should be separate from 'do build' action, but more on that in a bit.

> buildcache index wasn't taken into account

It was, a bit. I've found it only works in a pretty narrow set of circumstances, namely with a 'local filesystem' cache. I assume it works with S3 too. If you give it a web URL, it dies hard.  I wasn't sure what to expect, it's barely documented, and I haven't found a place where it's actually used. Fetching binary builds when spack installing seems to work fine without an index, and buildcache create doesn't use it either.

> What mental model should we have?

That you should always read from 'fetch', you should always write/delete from 'push', and that the URL scheme determines if push and fetch are strongly connected (ie, writing to 'push' directly effects 'fetch').

If a user wants the old behavior, they just have to NOT use the default URL scheme with separate push and fetch.  Ideally, I should create a new separate URL scheme for this behavior and the default URL scheme should check to make sure `push == fetch`. 

> What mental model should we _IDEALLY_ have?

That how to create the data for the binary cache and how to move it into place are entirely separate, and the URL scheme is responsible for the latter.  All URL schemes should implement 'fetch', but 'push' and (independently) 'delete' are separate and optional.

The URL scheme should provide:
   - A mechanism to determine validity of the mirror config. For example, if 'fetch must equal push`, then misconfigurations should give a useful error.
   - It's own authentication mechanisms and mirror config fields to manage them.
   - A mechanism to determine if a binary package _should_ be rebuilt.
   - An (optional) separate mechanism to push files.  
   - An (optional) separate mechanism (and URL) to delete files.
 
From the implementation side:
 - The URL scheme should probably be a class (or a variety of provided Mirror subclasses).
 - URL scheme operations will need the whole mirror definition, not just the push_url, and everything needed to operate should be in that definition.
 - It's a bad assumption that the path to the file and the URL are combined in a way that can be generalized.   
 - Auth parameters should be mirror specific.

Why all this?  Because the current setup supports local files, HTTP (read only), and S3, and literally nothing else. Almost all of these changes would be required to support the extremely simple HTTP based mechanisms I'm using, let alone a RESTful API, ftp, or anything else.

I will eventually get around to adding the HTTP PUT/DELETE based scheme I'm using to Spack, and all that entails. Right now I just rely on some scripts that use `curl` to get it done.

**Comment by becker33** [comment]
@Paul-Ferrell this sounds a lot like you're trying to recreate the logic of `spack ci rebuild` to use outside of the CI environment. Is it possible that that is a better approach to get this working? I'm not sure I fully understand the use case but it seems really similar to what we do in the gitlab pipelines.

**Comment by Paul-Ferrell** [comment]
@becker33 - Wow, I had to dig into the code quite a bit to figure out what `ci rebuild` actually does, which is a lot more than its documentation says. As to whether this would work for us, the answer at this point is: Maybe, but probably not. 

While `ci rebuild` will push to a separate local cache and only create binary packages that don't already exist in a separate binary cache location like we want, it starts stepping on other parts of our process that are well established at this point. We would love to see some of the features here integrated into the regular Spack commands. Namely, the ability to say 'hey, don't even build this when you do a `spack install` if it exists in this (or any) binary cache.', as well as the ability to look at one binarycache and write to another (which this PR tries to fix in its own way).

As our stuff is, our CI script basically just runs our production install script with a few extra options. This ensures that we're testing both the build process as well as the build configuration in one go.  To do it the `spack ci rebuild` way, we'd need to extract the full spec for each package we want to build from the concretized environment and feed them back to Spack. It's doable, but messy, and is certainly not a test of our production build process.

**Comment by Paul-Ferrell** [comment]
I guess I could do our normal `spack install`, and then iterate over the spec hashes like I already have do for `spack buildcache create`. The vast majority of `spack ci rebuild` would do nothing, as the build would would already be there. I wouldn't be able to use `spack ci generate`; it doesn't work with stacks (it has problems with matrix generated specs). We'd have to dynamically add both mirrors and the gitlab-ci section to the environment's spack.yaml; we try to keep any general spack configuration out of that file so that it can be dealt with on a per-site basis elsewhere.  It's a terrible fit, really, to do something that should be simple.

I'd much rather have this PR merged though, or a similar one that adds a second 'upstream-mirror' argument or something to `spack buildcache create`.

**Comment by alalazo** [comment]
What is the state of this PR? Can we close it?

**Comment by scottwittenburg** [comment]
> What is the state of this PR? Can we close it?

No objection from me.

**Comment by Paul-Ferrell** [comment]
There have been other merge requests that basically fix this for us.  Or, more precisely, seemed to fix this issue yet don't actually, and we've put through other merge requests to finish the job.  So yes, you can close it.

**Pull request closed without merging**
